oraclesrv/utils.py

```diff
@@ -142,3 +150,104 @@ def get_solr_data_match_thesis(author, year, doctype):
+        return True, 'updated db with a new record successfully'
+    except SQLAlchemyError as e:
+        current_app.logger.error('SQLAlchemy: ' + str(e))
+        return False, 'SQLAlchemy: ' + str(e)
+
+def get_a_record(source_bibcode, matched_bibcode):
+    """
+
+    :param source_bibcode:
+    :param matched_bibcode:
+    :return:
+    """
+    with current_app.session_scope() as session:
+        row = session.query(DocMatch).filter(or_(DocMatch.source_bibcode == source_bibcode,
+                                                 DocMatch.matched_bibcode == source_bibcode,
+                                                 DocMatch.source_bibcode == matched_bibcode,
+                                                 DocMatch.matched_bibcode == matched_bibcode)).order_by(desc(DocMatch.confidence)).first()
+        if row:
+            current_app.logger.debug("Fetched a record for matched bibcodes = (%s, %s)."  % (source_bibcode, matched_bibcode))
+            return row.toJSON()
+
+    current_app.logger.error("Failed to fetched a record for matched bibcodes = (%s, %s)."  % (source_bibcode, matched_bibcode))
+    return None
+
+def add_records(docmatches):
+    """
+    upserts records into db
+
+    :param docmatches:
+    :return: success boolean, plus a status text for retuning error message, if any, to the calling program
+    """
+    rows = []
+    for doc in docmatches.docmatch_records:
+        rows.append({"source_bibcode":doc.source_bibcode,
+                     "matched_bibcode": doc.matched_bibcode,
+                     "confidence": doc.confidence})
+
+    if len(rows) > 0:
+        table = DocMatch.__table__
+        stmt = insert(table).values(rows)
+
+        # get list of fields making up primary key
+        primary_keys = [c.name for c in list(table.primary_key.columns)]
+        # define dict of non-primary keys for updating
+        update_dict = {c.name: c for c in stmt.excluded if not c.primary_key}
+
+        on_conflict_stmt = stmt.on_conflict_do_update(index_elements=primary_keys, set_=update_dict)
+
+        try:
+            with current_app.session_scope() as session:
+                session.execute(on_conflict_stmt)
+            current_app.logger.info('updated db with new data successfully')
+            return True, 'updated db with new data successfully'
+        except SQLAlchemyError as e:
+            session.rollback()
+            current_app.logger.error('SQLAlchemy: ' + str(e))
+            return False, 'SQLAlchemy: ' + str(e)
+    return False, 'unable to add records to the database'
+
+def del_records(docmatches):
+    """
+    delete records from db
+
+    :param docmatches:
+    :return:
+    """
+    try:
+        with current_app.session_scope() as session:
+            count = 0
+            for doc in docmatches.docmatch_records:
+                count += session.query(DocMatch).filter(and_(DocMatch.source_bibcode == doc.source_bibcode,
```

Comment:
what happens if `delete()` is called when `filter()` finds no object?

Reply:
it returns 0. I have a unittest that attempt to delete the same records twice, in the second try it says 0 records has been deleted https://github.com/golnazads/oracle_service/blob/master/oraclesrv/tests/unittests/test_oracle_db.py#L95